// @elastic/elasticsearch/esm/api/index.js
/*
 * Copyright Elasticsearch B.V. and contributors
 * SPDX-License-Identifier: Apache-2.0
 */
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v, _w, _x, _y, _z, _0, _1, _2, _3, _4, _5, _6, _7, _8, _9, _10, _11, _12, _13, _14, _15, _16, _17, _18;
/* eslint-disable import/export */
/* eslint-disable @typescript-eslint/no-misused-new */
/* eslint-disable @typescript-eslint/no-extraneous-class */
/* eslint-disable @typescript-eslint/no-unused-vars */
// This file was automatically generated by elastic/elastic-client-generator-js
// DO NOT MODIFY IT BY HAND. Instead, modify the source open api file,
// and elastic/elastic-client-generator-js to regenerate this file again.
import AsyncSearchApi from './api/async_search.js';
import AutoscalingApi from './api/autoscaling.js';
import bulkApi from './api/bulk.js';
import capabilitiesApi from './api/capabilities.js';
import CatApi from './api/cat.js';
import CcrApi from './api/ccr.js';
import clearScrollApi from './api/clear_scroll.js';
import closePointInTimeApi from './api/close_point_in_time.js';
import ClusterApi from './api/cluster.js';
import ConnectorApi from './api/connector.js';
import countApi from './api/count.js';
import createApi from './api/create.js';
import DanglingIndicesApi from './api/dangling_indices.js';
import deleteApi from './api/delete.js';
import deleteByQueryApi from './api/delete_by_query.js';
import deleteByQueryRethrottleApi from './api/delete_by_query_rethrottle.js';
import deleteScriptApi from './api/delete_script.js';
import EnrichApi from './api/enrich.js';
import EqlApi from './api/eql.js';
import EsqlApi from './api/esql.js';
import existsApi from './api/exists.js';
import existsSourceApi from './api/exists_source.js';
import explainApi from './api/explain.js';
import FeaturesApi from './api/features.js';
import fieldCapsApi from './api/field_caps.js';
import FleetApi from './api/fleet.js';
import getApi from './api/get.js';
import getScriptApi from './api/get_script.js';
import getScriptContextApi from './api/get_script_context.js';
import getScriptLanguagesApi from './api/get_script_languages.js';
import getSourceApi from './api/get_source.js';
import GraphApi from './api/graph.js';
import healthReportApi from './api/health_report.js';
import IlmApi from './api/ilm.js';
import indexApi from './api/index.js';
import IndicesApi from './api/indices.js';
import InferenceApi from './api/inference.js';
import infoApi from './api/info.js';
import IngestApi from './api/ingest.js';
import knnSearchApi from './api/knn_search.js';
import LicenseApi from './api/license.js';
import LogstashApi from './api/logstash.js';
import mgetApi from './api/mget.js';
import MigrationApi from './api/migration.js';
import MlApi from './api/ml.js';
import MonitoringApi from './api/monitoring.js';
import msearchApi from './api/msearch.js';
import msearchTemplateApi from './api/msearch_template.js';
import mtermvectorsApi from './api/mtermvectors.js';
import NodesApi from './api/nodes.js';
import openPointInTimeApi from './api/open_point_in_time.js';
import pingApi from './api/ping.js';
import ProfilingApi from './api/profiling.js';
import ProjectApi from './api/project.js';
import putScriptApi from './api/put_script.js';
import QueryRulesApi from './api/query_rules.js';
import rankEvalApi from './api/rank_eval.js';
import reindexApi from './api/reindex.js';
import reindexRethrottleApi from './api/reindex_rethrottle.js';
import renderSearchTemplateApi from './api/render_search_template.js';
import RollupApi from './api/rollup.js';
import scriptsPainlessExecuteApi from './api/scripts_painless_execute.js';
import scrollApi from './api/scroll.js';
import searchApi from './api/search.js';
import SearchApplicationApi from './api/search_application.js';
import searchMvtApi from './api/search_mvt.js';
import searchShardsApi from './api/search_shards.js';
import searchTemplateApi from './api/search_template.js';
import SearchableSnapshotsApi from './api/searchable_snapshots.js';
import SecurityApi from './api/security.js';
import ShutdownApi from './api/shutdown.js';
import SimulateApi from './api/simulate.js';
import SlmApi from './api/slm.js';
import SnapshotApi from './api/snapshot.js';
import SqlApi from './api/sql.js';
import SslApi from './api/ssl.js';
import StreamsApi from './api/streams.js';
import SynonymsApi from './api/synonyms.js';
import TasksApi from './api/tasks.js';
import termsEnumApi from './api/terms_enum.js';
import termvectorsApi from './api/termvectors.js';
import TextStructureApi from './api/text_structure.js';
import TransformApi from './api/transform.js';
import updateApi from './api/update.js';
import updateByQueryApi from './api/update_by_query.js';
import updateByQueryRethrottleApi from './api/update_by_query_rethrottle.js';
import WatcherApi from './api/watcher.js';
import XpackApi from './api/xpack.js';
const kAsyncSearch = Symbol('AsyncSearch');
const kAutoscaling = Symbol('Autoscaling');
const kCat = Symbol('Cat');
const kCcr = Symbol('Ccr');
const kCluster = Symbol('Cluster');
const kConnector = Symbol('Connector');
const kDanglingIndices = Symbol('DanglingIndices');
const kEnrich = Symbol('Enrich');
const kEql = Symbol('Eql');
const kEsql = Symbol('Esql');
const kFeatures = Symbol('Features');
const kFleet = Symbol('Fleet');
const kGraph = Symbol('Graph');
const kIlm = Symbol('Ilm');
const kIndices = Symbol('Indices');
const kInference = Symbol('Inference');
const kIngest = Symbol('Ingest');
const kLicense = Symbol('License');
const kLogstash = Symbol('Logstash');
const kMigration = Symbol('Migration');
const kMl = Symbol('Ml');
const kMonitoring = Symbol('Monitoring');
const kNodes = Symbol('Nodes');
const kProfiling = Symbol('Profiling');
const kProject = Symbol('Project');
const kQueryRules = Symbol('QueryRules');
const kRollup = Symbol('Rollup');
const kSearchApplication = Symbol('SearchApplication');
const kSearchableSnapshots = Symbol('SearchableSnapshots');
const kSecurity = Symbol('Security');
const kShutdown = Symbol('Shutdown');
const kSimulate = Symbol('Simulate');
const kSlm = Symbol('Slm');
const kSnapshot = Symbol('Snapshot');
const kSql = Symbol('Sql');
const kSsl = Symbol('Ssl');
const kStreams = Symbol('Streams');
const kSynonyms = Symbol('Synonyms');
const kTasks = Symbol('Tasks');
const kTextStructure = Symbol('TextStructure');
const kTransform = Symbol('Transform');
const kWatcher = Symbol('Watcher');
const kXpack = Symbol('Xpack');
class API {
    constructor() {
        Object.defineProperty(this, _a, {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, _b, {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, _c, {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, _d, {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, _e, {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, _f, {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, _g, {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, _h, {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, _j, {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, _k, {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, _l, {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, _m, {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, _o, {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, _p, {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, _q, {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, _r, {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, _s, {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, _t, {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, _u, {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, _v, {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, _w, {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, _x, {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, _y, {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, _z, {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, _0, {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, _1, {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, _2, {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, _3, {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, _4, {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, _5, {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, _6, {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, _7, {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, _8, {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, _9, {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, _10, {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, _11, {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, _12, {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, _13, {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, _14, {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, _15, {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, _16, {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, _17, {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, _18, {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this[kAsyncSearch] = null;
        this[kAutoscaling] = null;
        this[kCat] = null;
        this[kCcr] = null;
        this[kCluster] = null;
        this[kConnector] = null;
        this[kDanglingIndices] = null;
        this[kEnrich] = null;
        this[kEql] = null;
        this[kEsql] = null;
        this[kFeatures] = null;
        this[kFleet] = null;
        this[kGraph] = null;
        this[kIlm] = null;
        this[kIndices] = null;
        this[kInference] = null;
        this[kIngest] = null;
        this[kLicense] = null;
        this[kLogstash] = null;
        this[kMigration] = null;
        this[kMl] = null;
        this[kMonitoring] = null;
        this[kNodes] = null;
        this[kProfiling] = null;
        this[kProject] = null;
        this[kQueryRules] = null;
        this[kRollup] = null;
        this[kSearchApplication] = null;
        this[kSearchableSnapshots] = null;
        this[kSecurity] = null;
        this[kShutdown] = null;
        this[kSimulate] = null;
        this[kSlm] = null;
        this[kSnapshot] = null;
        this[kSql] = null;
        this[kSsl] = null;
        this[kStreams] = null;
        this[kSynonyms] = null;
        this[kTasks] = null;
        this[kTextStructure] = null;
        this[kTransform] = null;
        this[kWatcher] = null;
        this[kXpack] = null;
    }
}
_a = kAsyncSearch, _b = kAutoscaling, _c = kCat, _d = kCcr, _e = kCluster, _f = kConnector, _g = kDanglingIndices, _h = kEnrich, _j = kEql, _k = kEsql, _l = kFeatures, _m = kFleet, _o = kGraph, _p = kIlm, _q = kIndices, _r = kInference, _s = kIngest, _t = kLicense, _u = kLogstash, _v = kMigration, _w = kMl, _x = kMonitoring, _y = kNodes, _z = kProfiling, _0 = kProject, _1 = kQueryRules, _2 = kRollup, _3 = kSearchApplication, _4 = kSearchableSnapshots, _5 = kSecurity, _6 = kShutdown, _7 = kSimulate, _8 = kSlm, _9 = kSnapshot, _10 = kSql, _11 = kSsl, _12 = kStreams, _13 = kSynonyms, _14 = kTasks, _15 = kTextStructure, _16 = kTransform, _17 = kWatcher, _18 = kXpack;
export default API;
API.prototype.bulk = bulkApi;
API.prototype.capabilities = capabilitiesApi;
API.prototype.clearScroll = clearScrollApi;
API.prototype.closePointInTime = closePointInTimeApi;
API.prototype.count = countApi;
API.prototype.create = createApi;
API.prototype.delete = deleteApi;
API.prototype.deleteByQuery = deleteByQueryApi;
API.prototype.deleteByQueryRethrottle = deleteByQueryRethrottleApi;
API.prototype.deleteScript = deleteScriptApi;
API.prototype.exists = existsApi;
API.prototype.existsSource = existsSourceApi;
API.prototype.explain = explainApi;
API.prototype.fieldCaps = fieldCapsApi;
API.prototype.get = getApi;
API.prototype.getScript = getScriptApi;
API.prototype.getScriptContext = getScriptContextApi;
API.prototype.getScriptLanguages = getScriptLanguagesApi;
API.prototype.getSource = getSourceApi;
API.prototype.healthReport = healthReportApi;
API.prototype.index = indexApi;
API.prototype.info = infoApi;
API.prototype.knnSearch = knnSearchApi;
API.prototype.mget = mgetApi;
API.prototype.msearch = msearchApi;
API.prototype.msearchTemplate = msearchTemplateApi;
API.prototype.mtermvectors = mtermvectorsApi;
API.prototype.openPointInTime = openPointInTimeApi;
API.prototype.ping = pingApi;
API.prototype.putScript = putScriptApi;
API.prototype.rankEval = rankEvalApi;
API.prototype.reindex = reindexApi;
API.prototype.reindexRethrottle = reindexRethrottleApi;
API.prototype.renderSearchTemplate = renderSearchTemplateApi;
API.prototype.scriptsPainlessExecute = scriptsPainlessExecuteApi;
API.prototype.scroll = scrollApi;
API.prototype.search = searchApi;
API.prototype.searchMvt = searchMvtApi;
API.prototype.searchShards = searchShardsApi;
API.prototype.searchTemplate = searchTemplateApi;
API.prototype.termsEnum = termsEnumApi;
API.prototype.termvectors = termvectorsApi;
API.prototype.update = updateApi;
API.prototype.updateByQuery = updateByQueryApi;
API.prototype.updateByQueryRethrottle = updateByQueryRethrottleApi;
Object.defineProperties(API.prototype, {
    asyncSearch: {
        get() { return this[kAsyncSearch] === null ? (this[kAsyncSearch] = new AsyncSearchApi(this.transport)) : this[kAsyncSearch]; }
    },
    autoscaling: {
        get() { return this[kAutoscaling] === null ? (this[kAutoscaling] = new AutoscalingApi(this.transport)) : this[kAutoscaling]; }
    },
    cat: {
        get() { return this[kCat] === null ? (this[kCat] = new CatApi(this.transport)) : this[kCat]; }
    },
    ccr: {
        get() { return this[kCcr] === null ? (this[kCcr] = new CcrApi(this.transport)) : this[kCcr]; }
    },
    cluster: {
        get() { return this[kCluster] === null ? (this[kCluster] = new ClusterApi(this.transport)) : this[kCluster]; }
    },
    connector: {
        get() { return this[kConnector] === null ? (this[kConnector] = new ConnectorApi(this.transport)) : this[kConnector]; }
    },
    danglingIndices: {
        get() { return this[kDanglingIndices] === null ? (this[kDanglingIndices] = new DanglingIndicesApi(this.transport)) : this[kDanglingIndices]; }
    },
    enrich: {
        get() { return this[kEnrich] === null ? (this[kEnrich] = new EnrichApi(this.transport)) : this[kEnrich]; }
    },
    eql: {
        get() { return this[kEql] === null ? (this[kEql] = new EqlApi(this.transport)) : this[kEql]; }
    },
    esql: {
        get() { return this[kEsql] === null ? (this[kEsql] = new EsqlApi(this.transport)) : this[kEsql]; }
    },
    features: {
        get() { return this[kFeatures] === null ? (this[kFeatures] = new FeaturesApi(this.transport)) : this[kFeatures]; }
    },
    fleet: {
        get() { return this[kFleet] === null ? (this[kFleet] = new FleetApi(this.transport)) : this[kFleet]; }
    },
    graph: {
        get() { return this[kGraph] === null ? (this[kGraph] = new GraphApi(this.transport)) : this[kGraph]; }
    },
    ilm: {
        get() { return this[kIlm] === null ? (this[kIlm] = new IlmApi(this.transport)) : this[kIlm]; }
    },
    indices: {
        get() { return this[kIndices] === null ? (this[kIndices] = new IndicesApi(this.transport)) : this[kIndices]; }
    },
    inference: {
        get() { return this[kInference] === null ? (this[kInference] = new InferenceApi(this.transport)) : this[kInference]; }
    },
    ingest: {
        get() { return this[kIngest] === null ? (this[kIngest] = new IngestApi(this.transport)) : this[kIngest]; }
    },
    license: {
        get() { return this[kLicense] === null ? (this[kLicense] = new LicenseApi(this.transport)) : this[kLicense]; }
    },
    logstash: {
        get() { return this[kLogstash] === null ? (this[kLogstash] = new LogstashApi(this.transport)) : this[kLogstash]; }
    },
    migration: {
        get() { return this[kMigration] === null ? (this[kMigration] = new MigrationApi(this.transport)) : this[kMigration]; }
    },
    ml: {
        get() { return this[kMl] === null ? (this[kMl] = new MlApi(this.transport)) : this[kMl]; }
    },
    monitoring: {
        get() { return this[kMonitoring] === null ? (this[kMonitoring] = new MonitoringApi(this.transport)) : this[kMonitoring]; }
    },
    nodes: {
        get() { return this[kNodes] === null ? (this[kNodes] = new NodesApi(this.transport)) : this[kNodes]; }
    },
    profiling: {
        get() { return this[kProfiling] === null ? (this[kProfiling] = new ProfilingApi(this.transport)) : this[kProfiling]; }
    },
    project: {
        get() { return this[kProject] === null ? (this[kProject] = new ProjectApi(this.transport)) : this[kProject]; }
    },
    queryRules: {
        get() { return this[kQueryRules] === null ? (this[kQueryRules] = new QueryRulesApi(this.transport)) : this[kQueryRules]; }
    },
    rollup: {
        get() { return this[kRollup] === null ? (this[kRollup] = new RollupApi(this.transport)) : this[kRollup]; }
    },
    searchApplication: {
        get() { return this[kSearchApplication] === null ? (this[kSearchApplication] = new SearchApplicationApi(this.transport)) : this[kSearchApplication]; }
    },
    searchableSnapshots: {
        get() { return this[kSearchableSnapshots] === null ? (this[kSearchableSnapshots] = new SearchableSnapshotsApi(this.transport)) : this[kSearchableSnapshots]; }
    },
    security: {
        get() { return this[kSecurity] === null ? (this[kSecurity] = new SecurityApi(this.transport)) : this[kSecurity]; }
    },
    shutdown: {
        get() { return this[kShutdown] === null ? (this[kShutdown] = new ShutdownApi(this.transport)) : this[kShutdown]; }
    },
    simulate: {
        get() { return this[kSimulate] === null ? (this[kSimulate] = new SimulateApi(this.transport)) : this[kSimulate]; }
    },
    slm: {
        get() { return this[kSlm] === null ? (this[kSlm] = new SlmApi(this.transport)) : this[kSlm]; }
    },
    snapshot: {
        get() { return this[kSnapshot] === null ? (this[kSnapshot] = new SnapshotApi(this.transport)) : this[kSnapshot]; }
    },
    sql: {
        get() { return this[kSql] === null ? (this[kSql] = new SqlApi(this.transport)) : this[kSql]; }
    },
    ssl: {
        get() { return this[kSsl] === null ? (this[kSsl] = new SslApi(this.transport)) : this[kSsl]; }
    },
    streams: {
        get() { return this[kStreams] === null ? (this[kStreams] = new StreamsApi(this.transport)) : this[kStreams]; }
    },
    synonyms: {
        get() { return this[kSynonyms] === null ? (this[kSynonyms] = new SynonymsApi(this.transport)) : this[kSynonyms]; }
    },
    tasks: {
        get() { return this[kTasks] === null ? (this[kTasks] = new TasksApi(this.transport)) : this[kTasks]; }
    },
    textStructure: {
        get() { return this[kTextStructure] === null ? (this[kTextStructure] = new TextStructureApi(this.transport)) : this[kTextStructure]; }
    },
    transform: {
        get() { return this[kTransform] === null ? (this[kTransform] = new TransformApi(this.transport)) : this[kTransform]; }
    },
    watcher: {
        get() { return this[kWatcher] === null ? (this[kWatcher] = new WatcherApi(this.transport)) : this[kWatcher]; }
    },
    xpack: {
        get() { return this[kXpack] === null ? (this[kXpack] = new XpackApi(this.transport)) : this[kXpack]; }
    }
});
