// @elastic/transport/lib/index.js
"use strict";
/*
 * Copyright Elasticsearch B.V. and contributors
 * SPDX-License-Identifier: Apache-2.0
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.events = exports.errors = exports.Serializer = exports.UndiciConnection = exports.HttpConnection = exports.BaseConnection = exports.CloudConnectionPool = exports.BaseConnectionPool = exports.ClusterConnectionPool = exports.WeightedConnectionPool = exports.Transport = exports.Diagnostic = void 0;
const tslib_1 = require("tslib");
const Diagnostic_1 = tslib_1.__importStar(require("./Diagnostic"));
exports.Diagnostic = Diagnostic_1.default;
Object.defineProperty(exports, "events", { enumerable: true, get: function () { return Diagnostic_1.events; } });
const Transport_1 = tslib_1.__importDefault(require("./Transport"));
exports.Transport = Transport_1.default;
const connection_1 = require("./connection");
Object.defineProperty(exports, "BaseConnection", { enumerable: true, get: function () { return connection_1.BaseConnection; } });
Object.defineProperty(exports, "HttpConnection", { enumerable: true, get: function () { return connection_1.HttpConnection; } });
Object.defineProperty(exports, "UndiciConnection", { enumerable: true, get: function () { return connection_1.UndiciConnection; } });
const pool_1 = require("./pool");
Object.defineProperty(exports, "WeightedConnectionPool", { enumerable: true, get: function () { return pool_1.WeightedConnectionPool; } });
Object.defineProperty(exports, "ClusterConnectionPool", { enumerable: true, get: function () { return pool_1.ClusterConnectionPool; } });
Object.defineProperty(exports, "CloudConnectionPool", { enumerable: true, get: function () { return pool_1.CloudConnectionPool; } });
Object.defineProperty(exports, "BaseConnectionPool", { enumerable: true, get: function () { return pool_1.BaseConnectionPool; } });
const Serializer_1 = tslib_1.__importDefault(require("./Serializer"));
exports.Serializer = Serializer_1.default;
const errors = tslib_1.__importStar(require("./errors"));
exports.errors = errors;
